research/fast-api/index.md

```diff
@@ -0,0 +1,246 @@
+                # from old data, sometimes build_logs_url is same and sometimes different to web_url
+                "build_logs_urls": build.build_logs_urls,
+                "pr_id": build.get_pr_id(),
+                "branch_name": build.get_branch_name(),
+                "release": build.get_release_tag(),
+            }
+
+            if project := build.get_project():
+                build_dict["project_url"] = project.project_url
+                build_dict["repo_namespace"] = project.namespace
+                build_dict["repo_name"] = project.repo_name
+
+            result.append(build_dict)
+
+        resp = response_maker(
+            result,
+            status=HTTPStatus.PARTIAL_CONTENT,
+        )
+        resp.headers["Content-Range"] = f"koji-builds {first + 1}-{last}/*"
+        return resp
+```
+
+### After (using Pydantic)
+
+```python
+from fastapi import APIRouter, Query, Request, Response, HTTPException
+from typing import List, Optional
+from pydantic import BaseModel
+from http import HTTPStatus
+
+router = APIRouter(prefix="/koji-builds", tags=["Production builds"])
+
+# Pydantic models
+class ProjectModel(BaseModel):
+    project_url: str
+    repo_namespace: str
+    repo_name: str
+
+
+class KojiBuildModel(BaseModel):
+    packit_id: int
+    task_id: int
+    scratch: bool
+    status: str
+    build_submitted_time: Optional[str]
+    chroot: str
+    web_url: str
+    build_logs_urls: List[str]
+    pr_id: Optional[int]
+    branch_name: Optional[str]
+    release: Optional[str]
+    project: Optional[ProjectModel] = None
+
+
+@router.get("", response_model=List[KojiBuildModel], status_code=HTTPStatus.PARTIAL_CONTENT)
+def get_koji_builds(
+    request: Request,
+    response: Response,
+    scratch: Optional[bool] = Query(None, description="Filter by scratch builds"),
+    first: int = Query(0, description="Pagination start index"),
+    last: int = Query(10, description="Pagination end index"),
+):
+    """List all Koji builds."""
+    result = []
+
+    for build in KojiBuildTargetModel.get_range(first, last, scratch):
+        build_data = KojiBuildModel(
+            packit_id=build.id,
+            task_id=build.task_id,
+            scratch=build.scratch,
+            status=build.status,
+            build_submitted_time=optional_timestamp(build.build_submitted_time),
+            chroot=build.target,
+            web_url=build.web_url,
+            build_logs_urls=build.build_logs_urls,
+            pr_id=build.get_pr_id(),
+            branch_name=build.get_branch_name(),
+            release=build.get_release_tag(),
+            project=(
+                ProjectModel(
+                    project_url=build.get_project().project_url,
+                    repo_namespace=build.get_project().namespace,
+                    repo_name=build.get_project().repo_name,
+                )
+                if build.get_project()
+                else None
+            ),
+        )
```

Comment:
This would also be moved or migrated to expected FastAPI way of doing things to make the endpoint self-documenting. Otherwise getting this documented over time is difficult without manual declaration. I think it should be possible in some sort of function wrapper to ensure the document is documented, or by using a pagination library together with this